Sources/Host/Darwin/Mach.cpp

```diff
@@ -80,14 +80,69 @@ ErrorCode Mach::readMemory(ProcessThreadId const &ptid, Address const &address,
   if (kret != KERN_SUCCESS)
     return kErrorInvalidAddress;
 
-  *count = curr_bytes_read;
+  if (count)
+    *count = curr_bytes_read;
 
   return kSuccess;
 }
 
 ErrorCode Mach::writeMemory(ProcessThreadId const &ptid, Address const &address,
                             void const *buffer, size_t length, size_t *count) {
-  return kErrorUnsupported;
+  /*
+   * We want to write in the process memory, but most of the write will be
+   * in the code section which will not be writable. So we need to do that
+   * in 4 step:
+   *
+   * 1) get current permissions
+   * 2) set write permissions
+   * 3) write
+   * 4) restore permissions
+   *
+   * TODO: Write in multipage ?
+   */
+  task_t task;
+
+  task = getMachTask(ptid.pid);
+  if (task == TASK_NULL) {
+    return kErrorProcessNotFound;
+  }
+
+  mach_msg_type_number_t infoCount = VM_REGION_SUBMAP_INFO_COUNT_64;
+  vm_region_submap_info_data_64_t regionInfo;
+  mach_vm_address_t start = address.value();
+  mach_vm_size_t size;
+  kern_return_t kret;
+  ErrorCode err = kSuccess;
+  natural_t depth = 1024;
+
+  kret =
+      mach_vm_region_recurse(task, &start, &size, &depth,
+                             (vm_region_recurse_info_t)&regionInfo, &infoCount);
+  if (kret != KERN_SUCCESS)
+    return kErrorInvalidAddress;
+
+  /*
+   * The combinason of VM_PROT_COPY/VM_PROT_READ is need to override the
+   * write protection.
+   */
+  kret = mach_vm_protect(task, address.value(), length, FALSE,
+                         VM_PROT_WRITE | VM_PROT_COPY | VM_PROT_READ);
+  if (kret != KERN_SUCCESS)
+    return kErrorInvalidAddress;
+
+  kret = mach_vm_write((vm_map_t)task, address.value(), (vm_offset_t)buffer,
+                       length);
+  if (kret != KERN_SUCCESS)
+    err = kErrorUnknown;
+  else if (count)
+    *count = length;
+
+  kret = mach_vm_protect(task, address.value(), length, FALSE,
+                         regionInfo.protection);
+  if (kret != KERN_SUCCESS)
```

Comment:
I think here you should see if `err` is already set to an error code (i.e.: if the previous call to `mach_vm_write` failed. Something like `if (kret != KERN_SUCCESS && err == kSuccess)` should be better, no?